### ISSUE TITLE
Query performance is twice that of java

### ISSUE THREAD

**Issue by travelliu** [godror/godror#250]
**Describe the bug**

Querying a table with more than 100 columns and traversing all the data takes twice as long as java.

how to improve performance ? 

**To Reproduce**
table 
```
CREATE TABLE "PUSHENG"."TAB_MANY_COLUMN" 
   (	"NE_KEY" NUMBER(19,0), 
	"STATISTIC_TIME" DATE, 
	"YEAR" NUMBER(4,0), 
	"MONTH" NUMBER(2,0), 
	"DAY" NUMBER(2,0), 
	"HOUR" NUMBER(2,0), 
	"INSERT_TIME" DATE, 
	"VENDOR_ID" NUMBER(4,0), 
	"COLL_SOURCE" NUMBER(10,0), 
	"COL_001" NUMBER(20,4), 
	"COL_002" NUMBER(20,4), 
	"COL_003" NUMBER(20,4), 
	"COL_004" NUMBER(20,4), 
	"COL_005" NUMBER(20,4), 
	"COL_006" NUMBER(20,4), 
	"COL_007" NUMBER(20,4), 
	"COL_008" NUMBER(20,4), 
	"COL_009" NUMBER(20,4), 
	"COL_010" NUMBER(20,4), 
	"COL_011" NUMBER(20,4), 
	"COL_012" NUMBER(20,4), 
	"COL_013" NUMBER(20,4), 
	"COL_014" NUMBER(20,4), 
	"COL_015" NUMBER(20,4), 
	"COL_016" NUMBER(20,4), 
	"COL_017" NUMBER(20,4), 
	"COL_018" NUMBER(20,4), 
	"COL_019" NUMBER(20,4), 
	"COL_020" NUMBER(20,4), 
	"COL_021" NUMBER(20,4), 
	"COL_022" NUMBER(20,4), 
	"COL_023" NUMBER(20,4), 
	"COL_024" NUMBER(20,4), 
	"COL_025" NUMBER(20,4), 
	"COL_026" NUMBER(20,4), 
	"COL_027" NUMBER(20,4), 
	"COL_028" NUMBER(20,4), 
	"COL_029" NUMBER(20,4), 
	"COL_030" NUMBER(20,4), 
	"COL_031" NUMBER(20,4), 
	"COL_032" NUMBER(20,4), 
	"COL_033" NUMBER(20,4), 
	"COL_034" NUMBER(20,4), 
	"COL_035" NUMBER(20,4), 
	"COL_036" NUMBER(20,4), 
	"COL_037" NUMBER(20,4), 
	"COL_038" NUMBER(20,4), 
	"COL_039" NUMBER(20,4), 
	"COL_040" NUMBER(20,4), 
	"COL_041" NUMBER(20,4), 
	"COL_042" NUMBER(20,4), 
	"COL_043" NUMBER(20,4), 
	"COL_044" NUMBER(20,4), 
	"COL_045" NUMBER(20,4), 
	"COL_046" NUMBER(20,4), 
	"COL_047" NUMBER(20,4), 
	"COL_048" NUMBER(20,4), 
	"COL_049" NUMBER(20,4), 
	"COL_050" NUMBER(20,4), 
	"COL_051" NUMBER(20,4), 
	"COL_052" NUMBER(20,4), 
	"COL_053" NUMBER(20,4), 
	"COL_054" NUMBER(20,4), 
	"COL_055" NUMBER(20,4), 
	"COL_056" NUMBER(20,4), 
	"COL_057" NUMBER(20,4), 
	"COL_058" NUMBER(20,4), 
	"COL_059" NUMBER(20,4), 
	"COL_060" NUMBER(20,4), 
	"COL_061" NUMBER(20,4), 
	"COL_062" NUMBER(20,4), 
	"COL_063" NUMBER(20,4), 
	"COL_064" NUMBER(20,4), 
	"COL_065" NUMBER(20,4), 
	"COL_066" NUMBER(20,4), 
	"COL_067" NUMBER(20,4), 
	"COL_068" NUMBER(20,4), 
	"COL_069" NUMBER(20,4), 
	"COL_070" NUMBER(20,4), 
	"COL_071" NUMBER(20,4), 
	"COL_072" NUMBER(20,4), 
	"COL_073" NUMBER(20,4), 
	"COL_074" NUMBER(20,4), 
	"COL_075" NUMBER(20,4), 
	"COL_076" NUMBER(20,4), 
	"COL_077" NUMBER(20,4), 
	"COL_078" NUMBER(20,4), 
	"COL_079" NUMBER(20,4), 
	"COL_080" NUMBER(20,4), 
	"COL_081" NUMBER(20,4), 
	"COL_082" NUMBER(20,4), 
	"COL_083" NUMBER(20,4), 
	"COL_084" NUMBER(20,4), 
	"COL_085" NUMBER(20,4), 
	"COL_086" NUMBER(20,4), 
	"COL_087" NUMBER(20,4), 
	"COL_088" NUMBER(20,4), 
	"COL_089" NUMBER(20,4), 
	"COL_090" NUMBER(20,4), 
	"COL_091" NUMBER(20,4), 
	"COL_092" NUMBER(20,4), 
	"COL_093" NUMBER(20,4), 
	"COL_094" NUMBER(20,4), 
	"COL_095" NUMBER(20,4), 
	"COL_096" NUMBER(20,4), 
	"COL_097" NUMBER(20,4), 
	"COL_098" NUMBER(20,4), 
	"COL_099" NUMBER(20,4), 
	"COL_100" NUMBER(20,4)
   ) SEGMENT CREATION IMMEDIATE 
  PCTFREE 10 PCTUSED 40 INITRANS 1 MAXTRANS 255 
 NOCOMPRESS LOGGING
  STORAGE(INITIAL 65536 NEXT 1048576 MINEXTENTS 1 MAXEXTENTS 2147483645
  PCTINCREASE 0 FREELISTS 1 FREELIST GROUPS 1
  BUFFER_POOL DEFAULT FLASH_CACHE DEFAULT CELL_FLASH_CACHE DEFAULT)
  TABLESPACE "USERS" ;
```

Java Code 

```
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class OracleQueryToCsv {
    private static final String CSV_FILE_NAME = "exported.csv";
    public static String host;
    public static String port;
    public static String user;
    public static String password;
    public static String database;
    public static Integer fetchSize;

    public static final SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss");

    public static void main(String[] args) throws ClassNotFoundException, IOException, SQLException {
//        initParameters(args);
        host = "172.16.0.198";
        port = "55446";
        database = "orcl";
        user = "xxxxx";
        password = "xxxxxx";
        fetchSize =2000;
        Class.forName("oracle.jdbc.OracleDriver");
//        Class.forName(oracle.jdbc.driver.OracleConnection.class.getName());
        String url = String.format("jdbc:oracle:thin:@//%s:%s/%s", host, port, database);

        PreparedStatement preparedStatement = null;
        ResultSet resultSet = null;
        File csvOutputFile = new File(CSV_FILE_NAME);
        try  {
            Connection oracleConnection = DriverManager.getConnection(url, user, password);
//            PrintWriter pw = new PrintWriter(csvOutputFile);
            preparedStatement = oracleConnection.prepareStatement(
                    "select NE_KEY, STATISTIC_TIME, YEAR, MONTH, DAY, HOUR, INSERT_TIME, VENDOR_ID, COLL_SOURCE, COL_001, COL_002, COL_003, COL_004, COL_005," +
                            " COL_006, COL_007, COL_008, COL_009, COL_010, COL_011, COL_012, COL_013, COL_014, COL_015, COL_016, COL_017, COL_018, COL_019, " +
                            "COL_020, COL_021, COL_022, COL_023, COL_024, COL_025, COL_026, COL_027, COL_028, COL_029, COL_030, COL_031, COL_032, COL_033, " +
                            "COL_034, COL_035, COL_036, COL_037, COL_038, COL_039, COL_040, COL_041, COL_042, COL_043, COL_044, COL_045, COL_046, COL_047, " +
                            "COL_048, COL_049, COL_050, COL_051, COL_052, COL_053, COL_054, COL_055, COL_056, COL_057, COL_058, COL_059, COL_060, COL_061, " +
                            "COL_062, COL_063, COL_064, COL_065, COL_066, COL_067, COL_068, COL_069, COL_070, COL_071, COL_072, COL_073, COL_074, COL_075, " +
                            "COL_076, COL_077, COL_078, COL_079, COL_080, COL_081, COL_082, COL_083, COL_084, COL_085, COL_086, COL_087, COL_088, COL_089, " +
                            "COL_090, COL_091, COL_092, COL_093, COL_094, COL_095, COL_096, COL_097, COL_098, COL_099, COL_100 from PUSHENG.TAB_MANY_COLUMN",
                    ResultSet.TYPE_FORWARD_ONLY,
                    ResultSet.CONCUR_READ_ONLY);
            preparedStatement.setFetchSize(fetchSize);
            resultSet = preparedStatement.executeQuery();
            int count = 0;
            System.out.println(count + " rows exported at: " + df.format(new Date()));
            while (resultSet.next()) {
                count++;
                Object[] row = new Object[109];
                row[0] = resultSet.getInt(1);
                row[1] = resultSet.getDate(2);
                row[2] = resultSet.getInt(3);
                row[3] = resultSet.getInt(4);
                row[4] = resultSet.getInt(5);
                row[5] = resultSet.getInt(6);
                row[6] = resultSet.getDate(7);
                for (int i = 8; i < 110; i++) {
                    row[i - 1] = resultSet.getInt(i);
                }
                // pw.println(convertToCSV(row));
                if (count % 100000 == 0) {
                    Thread.sleep(0);
                    System.out.println(count + " rows exported at: " + df.format(new Date()));
                }
            }
            System.out.println(count + " rows exported at: " + df.format(new Date()));
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
//            preparedStatement.close();
//            resultSet.close();
        }
    }

    public static String convertToCSV(Object[] data) {
        return Stream.of(data).map(each -> escapeSpecialCharacters(each.toString())).collect(Collectors.joining(","));
    }

    public static String escapeSpecialCharacters(String data) {
        String escapedData = data.replaceAll("\\R", " ");
        if (data.contains(",") || data.contains("\"") || data.contains("'")) {
            data = data.replace("\"", "\"\"");
            escapedData = "\"" + data + "\"";
        }
        return escapedData;
    }


    public static void initParameters(String[] args) throws IOException {
        if (args.length == 0) {
            System.out.println("需要指定 HOST ,PORT ,DATABASE, USER, PASSWORD");
            System.exit(0);
        }
        if (args.length == 1) {
            if (args[0].equals("--help")) {
                System.out.println("暂无");
                System.exit(0);
            }
        }

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--host":
                    if (i > args.length - 1) {
                        System.out.println("--host needs parameter!");
                        System.exit(0);
                    }
                    host = args[++i];
                    break;
                case "--port":
                    if (i > args.length - 1) {
                        System.out.println("--port needs parameter!");
                        System.exit(0);
                    }
                    port = args[++i];
                    break;
                case "--database":
                    if (i > args.length - 1) {
                        System.out.println("--database needs parameter!");
                        System.exit(0);
                    }
                    database = args[++i];
                    break;
                case "--user":
                    if (i > args.length - 1) {
                        System.out.println("--user needs parameter!");
                        System.exit(0);
                    }
                    user = args[++i];
                    break;
                case "--fetchSize":
                    if (i > args.length - 1) {
                        System.out.println("--fetchSize needs parameter!");
                        System.exit(0);
                    }
                    fetchSize = Integer.valueOf(args[++i]);
                    break;
                case "--password":
                    if (i > args.length - 1) {
                        System.out.println("--password needs parameter!");
                        System.exit(0);
                    }
                    password = args[++i];
                    break;
                default:
                    System.out.println("unrecognized config parameter: " + args[i] + "!");
                    System.exit(0);
            }
        }
    }
}
```
```
javac OracleQueryToCsv.java 
export CLASSPATH=./ojdbc8.jar:.
java OracleQueryToCsv
0 rows exported at: 2022-09-09 04:05:58
100000 rows exported at: 2022-09-09 04:06:00
200000 rows exported at: 2022-09-09 04:06:02
300000 rows exported at: 2022-09-09 04:06:04
400000 rows exported at: 2022-09-09 04:06:06
500000 rows exported at: 2022-09-09 04:06:07
600000 rows exported at: 2022-09-09 04:06:09
700000 rows exported at: 2022-09-09 04:06:11
800000 rows exported at: 2022-09-09 04:06:12
900000 rows exported at: 2022-09-09 04:06:14
1000000 rows exported at: 2022-09-09 04:06:16
1100000 rows exported at: 2022-09-09 04:06:18
1200000 rows exported at: 2022-09-09 04:06:20
1300000 rows exported at: 2022-09-09 04:06:21
1400000 rows exported at: 2022-09-09 04:06:23
1440704 rows exported at: 2022-09-09 04:06:24
```

Golang Code

```
package main

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/godror/godror"
	"github.com/godror/godror/dsn"
	_ "github.com/sijms/go-ora/v2"
	"time"
)

var (
	driverName = "godror"
	dbHost     = "172.16.0.198"
	dbPort     = "55446"
	dbUser     = "xxxxx"
	dbUserPwd  = "xxxxx"
	dbName     = "orcl"
	fetchSize  = 5000
)

func main() {
	var dbUrl string
	cfg := godror.ConnectionParams{
		CommonParams: dsn.CommonParams{},
		ConnParams:   dsn.ConnParams{},
		PoolParams:   dsn.PoolParams{},
		// NewPassword:          dsn.Password{},
		StandaloneConnection: true,
	}
	if cfg.Username == "" {
		cfg.Username = dbUser
	}
	if cfg.Password.Len() == 0 {
		cfg.Password = godror.NewPassword(dbUserPwd)
	}
	if cfg.ConnectString == "" {
		cfg.ConnectString = fmt.Sprintf(
			"(DESCRIPTION=(CONNECT_TIMEOUT = %v)(ADDRESS_LIST="+
				"(ADDRESS=(PROTOCOL=TCP)(HOST=%s)(PORT=%v)))(CONNECT_DATA=(SERVER=DEDICATED)(SERVICE_NAME=%s)))",
			30, dbHost, dbPort, dbName,
		)
	}
	dbUrl = cfg.StringWithPassword()
	fmt.Println("connStr: ", dbUrl)
	db, err := sql.Open(driverName, dbUrl)
	if err != nil {
		fmt.Println(err)
		return
	}
	var rowNum int64
	tx, err := db.Begin()
	if err != nil {
		fmt.Println(err)
		return
	}
	defer tx.Commit()
	sqlCmd := "select * from pusheng.tab_many_column"

	var (
		rows *sql.Rows
	)
	rows, err = tx.QueryContext(context.Background(), sqlCmd, godror.PrefetchCount(fetchSize), godror.FetchArraySize(fetchSize))

	if err != nil {
		fmt.Println(err)
		return
	}
	defer rows.Close()
	var (
		statisticTime, insertTime                                             time.Time
		neKey, year, month, day, hour, vendorId, collSource                   string
		col1, col2, col3, col4, col5, col6, col7, col8, col9, col10           int64
		col11, col12, col13, col14, col15, col16, col17, col18, col19, col20  int64
		col21, col22, col23, col24, col25, col26, col27, col28, col29, col30  int64
		col31, col32, col33, col34, col35, col36, col37, col38, col39, col40  int64
		col41, col42, col43, col44, col45, col46, col47, col48, col49, col50  int64
		col51, col52, col53, col54, col55, col56, col57, col58, col59, col60  int64
		col61, col62, col63, col64, col65, col66, col67, col68, col69, col70  int64
		col71, col72, col73, col74, col75, col76, col77, col78, col79, col80  int64
		col81, col82, col83, col84, col85, col86, col87, col88, col89, col90  int64
		col91, col92, col93, col94, col95, col96, col97, col98, col99, col100 int64
	)
	bTime := time.Now()
	fmt.Printf("%v rows exported %s \n", rowNum, time.Now().String())
	for rows.Next() {
		if err := rows.Scan(&neKey, &statisticTime, &year, &month, &day, &hour,
			&insertTime, &vendorId, &collSource,
			&col1, &col2, &col3, &col4, &col5, &col6, &col7, &col8, &col9, &col10,
			&col11, &col12, &col13, &col14, &col15, &col16, &col17, &col18, &col19, &col20,
			&col21, &col22, &col23, &col24, &col25, &col26, &col27, &col28, &col29, &col30,
			&col31, &col32, &col33, &col34, &col35, &col36, &col37, &col38, &col39, &col40,
			&col41, &col42, &col43, &col44, &col45, &col46, &col47, &col48, &col49, &col50,
			&col51, &col52, &col53, &col54, &col55, &col56, &col57, &col58, &col59, &col60,
			&col61, &col62, &col63, &col64, &col65, &col66, &col67, &col68, &col69, &col70,
			&col71, &col72, &col73, &col74, &col75, &col76, &col77, &col78, &col79, &col80,
			&col81, &col82, &col83, &col84, &col85, &col86, &col87, &col88, &col89, &col90,
			&col91, &col92, &col93, &col94, &col95, &col96, &col97, &col98, &col99, &col100,
		); err != nil {

		}
		rowNum++

		if rowNum%200000 == 0 {
			fmt.Printf("%v rows exported %s \n", rowNum, time.Now().String())
		}
	}
	fmt.Printf("%v rows exported %s \n", rowNum, time.Now().String())
	eTime := time.Now()
	fmt.Printf("fetchSize: %04v Rows: %v Time: %vms\n", fetchSize, rowNum, eTime.Sub(bTime).Milliseconds())
}
```

```
go build main.go 
./main 
connStr:  user=xxxx password=xxxxx connectString="(DESCRIPTION=(CONNECT_TIMEOUT = 30)(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)(HOST=172.16.0.198)(PORT=55446)))(CONNECT_DATA=(SERVER=DEDICATED)(SERVICE_NAME=orcl)))"
configDir= connectionClass= enableEvents=0 externalAuth=0 heterogeneousPool=0
libDir= newPassword= noTimezoneCheck=0 poolIncrement=0 poolMaxSessions=0 poolMinSessions=0
poolSessionMaxLifetime=0s poolSessionTimeout=0s poolWaitTimeout=0s prelim=0
standaloneConnection=1 sysasm=0 sysdba=0 sysoper=0 timezone=
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+08:00"=800) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
0 rows exported 2022-09-09 16:26:26.522990396 +0800 CST m=+0.108250636 
200000 rows exported 2022-09-09 16:26:36.829327925 +0800 CST m=+10.414588187 
400000 rows exported 2022-09-09 16:26:46.554248893 +0800 CST m=+20.139509153 
600000 rows exported 2022-09-09 16:26:56.140373408 +0800 CST m=+29.725633668 
800000 rows exported 2022-09-09 16:27:05.713997956 +0800 CST m=+39.299258215 
1000000 rows exported 2022-09-09 16:27:15.642944555 +0800 CST m=+49.228204821 
1200000 rows exported 2022-09-09 16:27:26.861533725 +0800 CST m=+60.446793985 
1400000 rows exported 2022-09-09 16:27:37.036869765 +0800 CST m=+70.622130038 
1440704 rows exported 2022-09-09 16:27:38.96510968 +0800 CST m=+72.550369960 
fetchSize: 5000 Rows: 1440704 Time: 72442ms
```

**Your oracle client version**
12.2.0.1.0

**Your godror version**
 v0.34.0

**Your go version**
1.17.13

**Your gcc version**
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)

**Machine (please complete the following information):**
 - OS: [e.g. win] Linux 
 - Architecture: [e.g. x86_64] x86_64
 - Version: [e.g. 10] CentOS 7.6

**Comment by tgulacsi** [comment]
```
go test -run=^$ -bench=Wide -cpuprofile c.pprof -memprofile m.pprof
BenchmarkSelectWideN20_4-8            38         126401593 ns/op
BenchmarkSelectWideN9-8               62          99549792 ns/op
```

NUMBER(20,4) is got as DPI_NATIVE_BYTE, so is converted to string.
If the column type fits into an int64 (NUMBER(19)  for example), then it is retrieved as DPI_NATIVE_TYPE_INT64 and is waaay faster - lot less allocation.

**Comment by tgulacsi** [comment]
![m20](https://user-images.githubusercontent.com/256074/189347789-51ef7856-9f26-4008-bc91-2f8615efbe50.svg)
![m9](https://user-images.githubusercontent.com/256074/189347794-6423e7eb-6605-44be-8836-dd3e4cccd786.svg)

**Comment by tgulacsi** [comment]
Scanning a NUMBER(20,4) column into int64 is a call for trouble!

**Comment by travelliu** [comment]
> 



> ```
> go test -run=^$ -bench=Wide -cpuprofile c.pprof -memprofile m.pprof
> BenchmarkSelectWideN20_4-8            38         126401593 ns/op
> BenchmarkSelectWideN9-8               62          99549792 ns/op
> ```
> 
> NUMBER(20,4) is got as DPI_NATIVE_BYTE, so is converted to string. If the column type fits into an int64 (NUMBER(19) for example), then it is retrieved as DPI_NATIVE_TYPE_INT64 and is waaay faster - lot less allocation.
NUMBER(20,4) was string before and changed to int64 in order to align with java code. There is not much change in performance

**Comment by tgulacsi** [comment]
What does this mean? What have you measured?

**Comment by travelliu** [comment]
NUMBER(20,4) scan to string
NUMBER(19) scan to int64.  But the performance is not much improved

**Comment by tgulacsi** [comment]
\c3246982354ec0a6eb9bc2fa318a1119214adeff
```
$ benchstat <(go test -run=^$ -bench=Wide -count=10)
name                time/op
SelectWide/9-4        24.1ms ±32%
SelectWide/18-4       24.8ms ±13%
SelectWide/19-4       52.2ms ±56%
SelectWide/20,4-4     40.7ms ±17%
SelectWide/20,4B-4    32.8ms ± 0%

name                speed
SelectWide/9-4      2.13MB/s ±25%
SelectWide/18-4     2.00MB/s ±22%
SelectWide/19-4     1.04MB/s ±40%
SelectWide/20,4-4   1.25MB/s ±15%
SelectWide/20,4B-4  1.54MB/s ± 0%
```

Though I cannot make it to return NATIVE_TYPE_INT64, so everything goes through as []byte, it seems that the column's data type does matter, and there is a big difference between NUMBER(18) and NUMBER(19).

**Comment by travelliu** [comment]
https://www.syntio.net/en/labs-musings/efficient-fetching-of-data-from-oracle-database-in-golang

Seems to be a CGO problem

**Comment by tgulacsi** [comment]
Yes, cgo does slow things down, but avoiding it would be a big undertaking (write a native Go driver for Oracle, like https://github.com/oracle/python-oracledb/ Thin mode).

But the benchmark is clear: using `NUMBER(18)` (which DOES fit into an int64) is twice as fast as using `NUMBER(20,4)`.
Casting it to a DOUBLE speeds it up too, but not that much.

**Comment by travelliu** [comment]
tks.
Are there any plans to implement [python-oracledb ](https://github.com/oracle/python-oracledb) thin mode in go?

**Comment by tgulacsi** [comment]
> tks. Are there any plans to implement [python-oracledb ](https://github.com/oracle/python-oracledb) thin mode in go?

I don't have time for it, don't know about the Oracle guys (@anthony-tuininga, @cjbj).
The lack of support for external authentication cools my temper, too.

**Comment by cjbj** [comment]
We are currently working in parallel on improvements to python-oracledb and on exploring a Node.js thin driver mode for node-oracledb.  We don't have immediate plans to look at, or contribute to, a Go thin driver.

@tgulacsi I assume that external auth is high on your list of needs for any driver? Is there anything else you prioritize?

**Comment by tgulacsi** [comment]
I'm only using Go, and I do need the external auth.
The lack of it in the Python thin driver suggests that it is a big undertaking.

After some meditation, I think this feature needs sophisticated knowledge of the low-level working, and it is security related,
so it is a delicate matter and I'm not surprised that has lower priority.

**Comment by cjbj** [comment]
There are lots of interesting projects to be done for Thin mode.  The list does make it clear how much good technology is captured in the Oracle Client libraries.

**Comment by anthony-tuininga** [comment]
The latest release of python-oracledb (https://cjones-oracle.medium.com/python-oracledb-1-1-for-oracle-db-supports-azure-oauth-2-0-tokens-ae6ad7579b0e) does have support for *one* flavor of external authentication -- namely token authentication. Which flavor do you need?

**Comment by tgulacsi** [comment]
I use the one that authenticates through Oracle Wallet. I bet it's a complicated legacy technology :-P

**Comment by anthony-tuininga** [comment]
Thanks, @tgulacsi. I haven't looked into that in great depth yet so not sure how complicated or how legacy. :-) Will discuss with @cjbj regarding priorities, of course!
